### ISSUE TITLE
Issue integrating multifield in boundaries

### ISSUE THREAD

**Issue by oriolcg** [gridap/Gridap.jl#614]
Hi @fverdugo, 

As mentioned in a private conversation, I am getting an error when integrating a quantity that depends on a component of a `MultiField` over a boundary. In what follows I attach the reproducer of the error. 

```julia
module tmp2
using Gridap

domain = (0,2,0,1)
partition = (2,2)
model = CartesianDiscreteModel(domain,partition)
const k = 2
reffeᵤ = ReferenceFE(lagrangian,VectorValue{2,Float64},k)
reffe_p = ReferenceFE(lagrangian,Float64,k-1)
V₀ = TestFESpace(model, reffeᵤ, conformity=:H1)
U = TrialFESpace(V₀)
Q = TestFESpace(model, reffe_p, conformity=:C0)
P = TrialFESpace(Q)
Y = MultiFieldFESpace([V₀,Q])
X = MultiFieldFESpace([U,P])

Ω = Triangulation(model)

degree = 2*k
xh₀ = interpolate_everywhere([VectorValue(0.0,1.0),1.0],X)

Γc = BoundaryTriangulation(model)
dΓc = Measure(Γc, degree)
nΓc = get_normal_vector(Γc)

function compute_coefficients(xh)
    uh, ph = xh
    F_drag, F_lift = ∑( ∫( nΓc⋅(2*ε(uh)) - ph*nΓc )dΓc )
    return F_drag, F_lift
end

for i in 0:1000
    CD, CL = compute_coefficients(xh₀)
    println(i," " ,CD, " ", CL)
end
end
```

The error that I get is the following:

```
caused by: DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 2 and 3")
Stacktrace:
 [1] _bcs1 at ./broadcast.jl:498 [inlined]
 [2] _bcs at ./broadcast.jl:495 [inlined]
 [3] broadcast_shape at ./broadcast.jl:489 [inlined]
 [4] return_cache(::Gridap.Arrays.Broadcasting{typeof(LinearAlgebra.dot)}, ::Array{Gridap.TensorValues.TensorValue{2,2,Float64,4},1}, ::Array{Gridap.TensorValues.TensorValue{2,2,Float64,4},1}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:193
 [5] return_value at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:184 [inlined]
 [6] return_value at /Users/oriol/Progs/Gridap/Gridap.jl/src/Fields/FieldArrays.jl:500 [inlined]
 [7] return_type at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:62 [inlined]
 [8] lazy_map(::Gridap.Fields.BroadcastingFieldOpMap{typeof(LinearAlgebra.dot)}, ::Gridap.Geometry.FaceCompressedVector{Array{Gridap.TensorValues.TensorValue{2,2,Float64,4},1},Gridap.Geometry.FaceToCellGlue{Array{Int64,1},FillArrays.Fill{Int64,1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}},FillArrays.Fill{Int8,1,Tuple{Base.OneTo{Int64}}}}}, ::Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Fields.LinearCombinationMap{Colon},1,Tuple{Base.OneTo{Int64}}},Array{Gridap.TensorValues.TensorValue{2,2,Float64,4},1},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{Gridap.Arrays.PosNegReindex{Gridap.Arrays.SubVector{Float64,Array{Float64,1}},Array{Float64,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Float64,1},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Gridap.Arrays.Table{Int32,Array{Int32,1},Array{Int32,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Int32,1},1,Tuple{Array{Int32,1}}}}},Gridap.Geometry.FaceCompressedVector{Array{Gridap.TensorValues.TensorValue{2,2,Float64,4},2},Gridap.Geometry.FaceToCellGlue{Array{Int64,1},FillArrays.Fill{Int64,1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}},FillArrays.Fill{Int8,1,Tuple{Base.OneTo{Int64}}}}}}}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/LazyArrays.jl:57
 [9] lazy_map(::typeof(Gridap.Arrays.evaluate), ::Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{Gridap.Arrays.Operation{typeof(LinearAlgebra.dot)}},1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.OperationField{typeof(LinearAlgebra.dot),Tuple{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}}}},1,Tuple{FillArrays.Fill{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{typeof(Gridap.Fields.linear_combination),1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{Gridap.Arrays.PosNegReindex{Gridap.Arrays.SubVector{Float64,Array{Float64,1}},Array{Float64,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Float64,1},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Gridap.Arrays.Table{Int32,Array{Int32,1},Array{Int32,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Int32,1},1,Tuple{Array{Int32,1}}}}},Gridap.Fields.MemoArray{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}},1,FillArrays.Fill{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}},1,Tuple{Base.OneTo{Int64}}}}}}}}, ::Gridap.Geometry.FaceCompressedVector{Array{Gridap.TensorValues.VectorValue{2,Float64},1},Gridap.Geometry.FaceToCellGlue{Array{Int64,1},FillArrays.Fill{Int64,1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}},FillArrays.Fill{Int8,1,Tuple{Base.OneTo{Int64}}}}}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/Fields/ApplyOptimizations.jl:92
 [10] lazy_map at /Users/oriol/Progs/Gridap/Gridap.jl/src/Fields/ApplyOptimizations.jl:63 [inlined]
 [11] lazy_map(::typeof(Gridap.Arrays.evaluate), ::Gridap.Fields.MemoArray{Gridap.Fields.OperationField{Gridap.Fields.OperationField{typeof(LinearAlgebra.dot),Tuple{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}}}},Tuple{Gridap.Fields.LinearCombinationField{Array{Gridap.TensorValues.VectorValue{2,Float64},1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Polynomials.MonomialBasis{1,Float64}}}}},1,Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{typeof(∘)},1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.OperationField{Gridap.Fields.OperationField{typeof(LinearAlgebra.dot),Tuple{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}}}},Tuple{Gridap.Fields.LinearCombinationField{Array{Gridap.TensorValues.VectorValue{2,Float64},1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Polynomials.MonomialBasis{1,Float64}}}}},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{Gridap.Arrays.Operation{typeof(LinearAlgebra.dot)}},1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.OperationField{typeof(LinearAlgebra.dot),Tuple{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}}}},1,Tuple{FillArrays.Fill{Gridap.Fields.OperationField{typeof(Gridap.Fields.pinvJt),Tuple{Gridap.Fields.ConstantField{Gridap.TensorValues.TensorValue{2,2,Float64,4}}}},1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{typeof(Gridap.Fields.linear_combination),1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.LinearCombinationField{Array{Float64,1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}}},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Broadcasting{Gridap.Arrays.PosNegReindex{Gridap.Arrays.SubVector{Float64,Array{Float64,1}},Array{Float64,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Float64,1},1,Tuple{Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Gridap.Arrays.Table{Int32,Array{Int32,1},Array{Int32,1}}},1,Tuple{Base.OneTo{Int64}}},Array{Int32,1},1,Tuple{Array{Int32,1}}}}},Gridap.Fields.MemoArray{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}},1,FillArrays.Fill{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Fields.FieldGradientArray{1,Gridap.Polynomials.MonomialBasis{2,Gridap.TensorValues.VectorValue{2,Float64}},Gridap.Fields.FieldGradient{1,Gridap.Polynomials.Monomial},1}},1,Tuple{Base.OneTo{Int64}}}}}}}},Gridap.Arrays.LazyArray{FillArrays.Fill{typeof(Gridap.Fields.linear_combination),1,Tuple{Base.OneTo{Int64}}},Gridap.Fields.LinearCombinationField{Array{Gridap.TensorValues.VectorValue{2,Float64},1},Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Polynomials.MonomialBasis{1,Float64}}},1,Tuple{Gridap.Geometry.FaceCompressedVector{Array{Gridap.TensorValues.VectorValue{2,Float64},1},Gridap.Geometry.FaceToCellGlue{Array{Int64,1},FillArrays.Fill{Int64,1,Tuple{Base.OneTo{Int64}}},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}},FillArrays.Fill{Int8,1,Tuple{Base.OneTo{Int64}}}}},Gridap.Arrays.CompressedArray{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Polynomials.MonomialBasis{1,Float64}},1,Array{Gridap.Fields.LinearCombinationFieldVector{Array{Float64,2},Gridap.Polynomials.MonomialBasis{1,Float64}},1},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}}}}}}}}, ::Gridap.Arrays.CompressedArray{Array{Gridap.TensorValues.VectorValue{1,Float64},1},1,Array{Array{Gridap.TensorValues.VectorValue{1,Float64},1},1},Gridap.Arrays.LazyArray{FillArrays.Fill{Gridap.Arrays.Reindex{Array{Int8,1}},1,Tuple{Base.OneTo{Int64}}},Int8,1,Tuple{Array{Int64,1}}}}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/Fields/ApplyOptimizations.jl:319
 [12] evaluate!(::Nothing, ::Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain}, ::Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:390
 [13] evaluate at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:87 [inlined]
 [14] CellField at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:384 [inlined]
 [15] #22 at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:558 [inlined]
 [16] map(::Gridap.CellData.var"#22#23"{Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}}, ::Tuple{Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain}}) at ./tuple.jl:157
 [17] evaluate!(::Nothing, ::Gridap.CellData.OperationCellField{Gridap.CellData.ReferenceDomain}, ::Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:558
 [18] evaluate at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:87 [inlined]
 [19] CellField at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:384 [inlined]
 [20] #22 at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:558 [inlined]
 [21] map(::Gridap.CellData.var"#22#23"{Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}}, ::Tuple{Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain},Gridap.CellData.OperationCellField{Gridap.CellData.ReferenceDomain}}) at ./tuple.jl:158
 [22] evaluate!(::Nothing, ::Gridap.CellData.OperationCellField{Gridap.CellData.ReferenceDomain}, ::Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:558
 [23] evaluate at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:87 [inlined]
 [24] CellField at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:384 [inlined]
 [25] #14 at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:493 [inlined]
 [26] map(::Gridap.CellData.var"#14#17"{Gridap.CellData.CellPoint{Gridap.CellData.ReferenceDomain}}, ::Tuple{Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain},Gridap.CellData.OperationCellField{Gridap.CellData.ReferenceDomain}}) at ./tuple.jl:158
 [27] Gridap.CellData.OperationCellField(::Gridap.Arrays.Operation{typeof(LinearAlgebra.dot)}, ::Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain}, ::Vararg{Gridap.CellData.CellField,N} where N) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:493
 [28] _operate_cellfields(::Gridap.Arrays.Operation{typeof(LinearAlgebra.dot)}, ::Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain}, ::Vararg{Any,N} where N) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:569
 [29] evaluate! at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:454 [inlined]
 [30] evaluate at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:87 [inlined]
 [31] Map at /Users/oriol/Progs/Gridap/Gridap.jl/src/Arrays/Maps.jl:90 [inlined]
 [32] dot(::Gridap.CellData.GenericCellField{Gridap.CellData.ReferenceDomain}, ::Gridap.CellData.OperationCellField{Gridap.CellData.ReferenceDomain}) at /Users/oriol/Progs/Gridap/Gridap.jl/src/CellData/CellFields.jl:649
 [33] compute_coefficients(::Gridap.MultiField.MultiFieldFEFunction{Gridap.MultiField.MultiFieldCellField{Gridap.CellData.ReferenceDomain}}) at /Users/oriol/Progs/Gridap/Gridap.jl/tmp2.jl:28
 [34] top-level scope at /Users/oriol/Progs/Gridap/Gridap.jl/tmp2.jl:33
```

Note that the error appears at different iterations of the `for` loop (without any apparent patern). The most intriguing thing is that when I define the integration measure to be `dΓc = Measure(Γc, k)` instead of `dΓc = Measure(Γc, degree)` (with `k=2` and `degree=2*k`), the problem is solved. This makes me think that the problem is at the integration (or evaluation of the quantities at the quadrature points).

If I understand it correctly, it would be sufficient to use a degree in the `Measure` equal to `k` (polynomial order) because the solution would be already integrated exactly. But I don't see any conceptual problem on using a higher degree than the required.

**Comment by oriolcg** [comment]
Just an update on this issue. Exactly the same error appears in the `DGHeatEquationTests.jl` in `GridapODEs`, see https://github.com/gridap/GridapODEs.jl/runs/2788676141#step:6:481

This may indicate that the problem appears when evaluating boundary integrals. Could it be related to some of the optimizations done in latest Gridap version? I'm thinking on the line that some arrays are stored and reused in a later stage, but the integration points don't match.

**Comment by fverdugo** [comment]
>  Could it be related to some of the optimizations done in latest Gridap version?

Optimizations are the root of all evil. Thus, yes it could be.

I have notieced that the error happens in a different step `i` each time I execute the example. This is very strange!

**Comment by fverdugo** [comment]
It sould be fixed by PR #616. If not, please reopen the issue @oriolcg 

Closing for now

**Comment by fverdugo** [comment]
@oriolcg can you confirm that this has fixed the problem in your use cases?

if so, I'll will release the fix.
